### PR TITLE
Adding a basic no-media call that uses promises.

### PR DESCRIPTION
**Pull request by alvestrand**
This tests the basic promise machinery in PeerConnection.

Tested and works in Firefox; asking @jan-ivar to take a look because he's probably most passionate about how Promises usage should be styled.

### CODE REVIEW

**Comment by alvestrand** [comment]
Tagging @dontcallmedom

**Comment by alvestrand** [comment]
PTAL. I used a long then-chain with a terminating "catch", and a variable global to the chain to keep track of which step in the chain went "boom". (I dislike trying to use line numbers for this.)

It seems to me that with long promise-chains, per-step error handling is a Bad Idea, because the error handler can't tell whether it got its error from the function it's guarding or a previous step - and there's no way to abort the chain and cause something else to happen, given that throwing will just get you the next step's error handler.

In this case, a .catch at the end seems fit-for-purpose.

**Comment by jan-ivar** [comment]
Otherwise looks great!

**Comment by alvestrand** [comment]
Somehow your comments failed to attach to lines, which renders them a bit hard to parse.... where is step-naming different?
The "first" and "second" could be "caller" and "callee", but that's actually hard to distinguish visually. I inherited this from where I started working.
The reason for passing the offer/answer SDP through a global variable is that the implementation is allowed to modify the descriptions based on candidate generation, and I did not want that side effect to randomize what happened during this test (although it SHOULD have no effect).
My thought when adding the "atStep = negotiation completed" was that if something went wrong, I wanted to make sure the atStep did not show "setRemoteDescription first" after that step had completed. If debugging the script, one can look at the value :-)
Am going to submit a new update to take care of the whitespace error.

**Comment by jan-ivar** [comment]
I get lines [here](https://github.com/alvestrand/web-platform-tests/commit/698efbcd7b72b59bd7a83ed646d124b3a33964bf#commitcomment-11241364).

**Comment by jan-ivar** [comment]
> The reason for passing the offer/answer SDP through a global variable is that the implementation is allowed to modify the descriptions based on candidate generation, and I did not want that side effect to randomize what happened during this test (although it SHOULD have no effect).

This concerns me. Are people expected to know this? If candidates are generated, what's wrong with including them in the SDP sent to the peer (other than perhaps a redundant onicecandidate)? Is an implementation allowed to modify the SDP in other ways? If so, shouldn't the spec give clearer guidance here? If half the people send the pre-setLocalDescription description and the other half send the post-setLocalDescription description then that seems like the worst possible situation.

**Comment by alvestrand** [comment]
Cullen has authored a PR that somewhat clarifies what's contained in the localdescription and remotedescription: https://github.com/w3c/webrtc-pc/pull/225

I don't think it matters which description is sent; with trickling of candidates (which seems to be the default here) the connection is eventually made no matter what, but this is a test script. One of the things one wants to do with test scripts is to minimize variation for the parts one is not testing.

**Comment by jan-ivar** [comment]
Thanks for the link.

**Comment by dontcallmedom** [comment]
@jan-ivar can you confirm this is good to go from your perspective?

**Comment by jan-ivar** [comment]
I like tests that take the same code-paths that users would take. I see Harald's point in theory, but pragmatically, if the test ends up avoiding some race by being pedantic, then I see that as a disadvantage, not an advantage.

Also, to me, this pattern looks like a hold-over from the days when people sent offers from createOffer, something we now frown on since it's easy to check for failure on setLocalDescription with promises. I never did that in my tests, which is probably why the caching of the offer from createOffer (which excludes ice candidates that may have arrived in the time passed) seems so odd to me.

I don't want to belabor the point much further, as I don't know the scope of these tests well, whether they are meant to inform as examples or merely serve internal functions.

**Comment by jan-ivar** [comment]
Also, `onicecandidate` is not supposed to fire before `setLocalDescription` has finished anyway, so there can't be a difference here, so the caching seems redundant.

**Comment by jan-ivar** [comment]
Sorry for the wishy washy comments. I would like to see the test not cache offer and answer.

File: webrtc/promises-call.html [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  <script src="/resources/testharnessreport.js"></script>
+  <script src="/common/vendor-prefix.js"
+          data-prefixed-objects=
+              '[{"ancestors":["window"], "name":"RTCPeerConnection"},
+                {"ancestors":["window"], "name":"RTCSessionDescription"},
+                {"ancestors":["window"], "name":"RTCIceCandidate"}]'
+    >
+  </script>
+  <script type="text/javascript">
+  var test = async_test('Can set up a basic WebRTC call with no data using promises.');
```
Comment (dontcallmedom) [inline webrtc/promises-call.html]: "with no data"? or "with only data"?

Comment (alvestrand) [inline webrtc/promises-call.html]: only data. Fixed.

**Comment by alvestrand** [comment]
OK, I removed the caching of offer and answer; after sleeping on it, I couldn't convince myself it was worthwhile.

But then the explicit sequencing of setLocalDescription and setRemoteDescription started bothering me - so I removed that too, using Promise.all. WDYT?

This removes the ability to see which of first.setLocalDescription() and second.setRemoteDescription failed, but it strips down the example even more.

WDYT?

**Comment by jan-ivar** [comment]
This is a variation of signaling after createOffer, with the same problems I outlined earlier. By not waiting for setLocalDescription to verify the offer, there's error fan-out. i.e. if setLocalDescription fails for any reason then the signaling is already on its way to the peer, creating more failures than seems neecssary.

Also, I consider any delaying of setLocalDescription callback to be a footgun. It's the one method that is extremely time-sensitive in our API. Any guarantees about ice candidates not firing until after the setLocalDescription callback fires are lost here by waiting for both setLocal and setRemote to finish.

I don't think we should worry about the time it takes setLocalDescription to finish.

**Comment by alvestrand** [comment]
With respect, I think this argument is bogus.

First of all, this is a *platform test*.

The API description *guarantees* that in the callback from createOffer, the (unmodified) offer will be useful: "Session descriptions generated by createOffer must be immediately usable by setLocalDescription without causing an error as long as setLocalDescription is called reasonably soon."

If SetLocalDescription fails, there is a bug in the platform, and the test will fail. That's the whole point of the test.

Second, without knowing how promise.all is implemented internally, I don't see how this construct constitues a "delay of setLocalDescription". The result of setLocalDescription() will feed into Promise.all when setLocalDescription() is finished. If it fails, Promise.all will fail at once.

Third, I don't understand the time sensitivity you are claiming. One of the ends of the connection will start firing candidates before the other is ready; the other one won't - one of the endpoints' first probe will be lost, and *that is what we expect to happen*. They will connect. That's why ICE is probed from both ends.

Fourth, I don't even understand what guarantee you're claiming. Please cite the section in the spec that gives a guarantee, and tell me what it says.

**Comment by jan-ivar** [comment]
> The API description guarantees that in the callback from createOffer, the (unmodified) offer will be useful: 

... by setLocalDescription, not signaling.

Also, does this mean setLocalDescription MUST NOT fail if the description is unmodified? As written, it seems to me it could still fail for other reasons.

In any case, I now better understand why older tests and examples didn't bother checking for setLocalDescription failing, [but we have stamped out this behavior](http://w3c.github.io/webrtc-pc/#examples-and-call-flows) because we agreed it's a bad template (as A. it precludes other failures, and B. if you start modifying the sdp for some reason then - oops! - you have no error checking).

I don't think official platform tests should be exempt from this best practice without good reason, as people copy JS more than they read specs.

> One of the ends of the connection will start firing candidates before the other is ready; the other one won't - one of the endpoints' first probe will be lost, and that is what we expect to happen. They will connect. That's why ICE is probed from both ends.

This sounds like a bug. No candidates should fail. Here's a [Firefox (specific) test](http://jsfiddle.net/aynr0k5q) that succeeds without any candidate failures.

About the timing issue, ICE gathering starts asynchronously inside setLocalDescription [here](http://w3c.github.io/webrtc-pc/#widl-RTCPeerConnection-setLocalDescription-Promise-void--RTCSessionDescription-description):
 > 5. If the local description was set, connection's ice gathering state is new, and the local description contains media, then set connection's ice gathering state to gathering.

JSEP effectively prohibits onicecandidate from being called prior to the setLocalDescription callback, because a [naive web application](http://w3c.github.io/webrtc-pc/#examples-and-call-flows) would otherwise send its candidates before it sends its offer, which [JSEP prohibits](https://tools.ietf.org/html/draft-ietf-rtcweb-jsep-00#section-4.5).

Now, the test here is "cheating" by sending the offer before setLocalDescription, so it does not have this problem, whereas someone who would need to munge the SDP would not have this luxury. How relevant then is this test?

Imagine you need to munge the SDP, and move your gSecondConnection.setRemoteDescription() call to after setLocalDescription succeeds. Now put in a timing delay before right before it and watch what happens. You'll see candidates firing before the offer, and on gSecondConnection you'll see addIceCandidate happening before setRemoteDescription() which [should fail](https://code.google.com/p/webrtc/issues/detail?id=4409#c3).

**Comment by alvestrand** [comment]
Jan-Ivar, your thinking is muddy.

Re candidate probes failing: You'll notice this particular little piece of the log from your jsfiddle:

ICE-PEER(PC:1431938029319833 (id=21 url=http://localhost:8000/webrtc/promises-call.html?usePrefixes=1):default)/CAND-PAIR(REBC): triggered check on REBC|IP4:192.168.1.17:52671/UDP|IP4:192.168.1.17:46491/UDP(host(IP4:192.168.1.17:52671/UDP)|prflx)
ICE-PEER(PC:1431938029381961 (id=21 url=http://localhost:8000/webrtc/promises-call.html?usePrefixes=1):default)/CAND-PAIR(vC9l): triggered check on vC9l|IP4:192.168.1.17:46491/UDP|IP4:192.168.1.17:52671/UDP(host(IP4:192.168.1.17:46491/UDP)|candidate:0 1 UDP 2122252543 192.168.1.17 52671 typ host)
ICE-PEER(PC:1431938029319833 (id=21 url=http://localhost:8000/webrtc/promises-call.html?usePrefixes=1):default)/CAND-PAIR(REBC): triggered check on REBC|IP4:192.168.1.17:52671/UDP|IP4:192.168.1.17:46491/UDP(host(IP4:192.168.1.17:52671/UDP)|prflx)
ICE-PEER(PC:1431938029381961 (id=21 url=http://localhost:8000/webrtc/promises-call.html?usePrefixes=1):default)/CAND-PAIR(vC9l): triggered check on vC9l|IP4:192.168.1.17:46491/UDP|IP4:192.168.1.17:52671/UDP(host(IP4:192.168.1.17:46491/UDP)|candidate:0 1 UDP 2122252543 192.168.1.17 52671 typ host)
ICE-PEER(PC:1431938029319833 (id=21 url=http://localhost:8000/webrtc/promises-call.html?usePrefixes=1):default)/CAND-PAIR(REBC): setting pair to state SUCCEEDED: REBC|IP4:192.168.1.17:52671/UDP|IP4:192.168.1.17:46491/UDP(host(IP4:192.168.1.17:52671/UDP)|prflx)

How many probes were sent?
(You don't include timestamps in your log, so it's hard to see the timing of what happened here.)

You say "JSEP effectively prohibits onicecandidate from being called prior to the setLocalDescription callback, because a naive web application would otherwise send its candidates before it sends its offer, which JSEP prohibits." - if you want the API to have this restriction, then make that part of the spec. And JSEP doesn't prohibit what you say it prohibits - JSEP doesn't care - and *should* not care - how or when the application chooses to send its data.

What might make sense to prohibit is to call AddIceCandidate if there is no remoteDescription. And lo and behold, there is a restriction here, but not this one: "If this RTCPeerConnection object's signaling state is closed, the user agent must reject p with InvalidStateError, and jump to the step labeled Return."

Summary: The spec doesn't say what you claim it says. If you think it should say what you claim it says, file a bug. Until then, a test that exercises a legal sequence of transitions is a valid test of the current specification.

All that yelled .... I'll still switch away from using Promise.all in this test. Not because it's in any way, shape or form illegal, immoral or hazardous, but because the test could be copied into places where it might matter.

**Comment by jan-ivar** [comment]
I linked to a bug already https://code.google.com/p/webrtc/issues/detail?id=4409#c3

**Comment by jan-ivar** [comment]
About candidates, no need to look in logs. [Here's a fiddle](http://jsfiddle.net/3b2hfanw) that dumps them. If you un-comment the commented-out line you'll see addIce failures (proving that errors are not happening otherwise).

On JSEP, here's what ekr says: https://bugzilla.mozilla.org/show_bug.cgi?id=1087551#c9

**Comment by jan-ivar** [comment]
To quote JSEP:
> Candidate trickling is a technique through which a caller may incrementally provide candidates to the callee after the initial offer has been dispatched.

That seems clear to me. i.e. order matters. Otherwise re-negotiation becomes hard:

If Alice signals "offer", "candidate", "candidate", then Bob knows what that means.
If Alice signals "candidate", "offer", "candidate", then should Bob ignore the 1st candidate?

Does it belong to the new offer or a previous offer? My reading of JSEP is that order matters, and UAs should therefore not fire onnegotiationneeded so early we trick apps into misfiring.

**Comment by alvestrand** [comment]
@jan-ivar seems I forgot to push this fix before the weekend. Please take another look!

**Comment by jan-ivar** [comment]
lgtm.
